### PR TITLE
[ios][home] Add three finger gesture to open the dev menu + some refactoring

### LINKED ISSUE #7644: Touch Gesture for Dev Menu in Expo Client
### Background 

We removed the two finger touch gesture option to bring up the dev menu in the latest version of the client. We are planning to add a developer tools keyboard shortcut to open the dev menu but we will probably have to wait for RN 0.62 for that.

One reason we removed this was just to simplify things but shaking your phone is really annoying. There are keyboard shortcuts you can use on the simulator, but many users are on Windows or Linux machines (me!) and so they can't even use the iOS simulator and the Android emulator is still not great unless you have an extremely powerful computer. There are also lots of other good reasons to develop on your device.

A number of users reported this as an issue or otherwise complained about it. Ex. https://github.com/expo/expo/issues/7629 and others.

### Proposal

I think we should add a new gesture that can trigger the dev menu -- a three finger touch. It's pretty common to use two finger touch in an app -- if you have any pinch or zoom interactions, for example -- and so that often can conflict with the old two finger gesture. Almost no app interactions use three finger touches for anything, so it would be safe in almost any context.

Instead of making the settings let you pick between shake and multi finger touch like we did before, I think we should just have two checkbox-type settings, both enabled by default. That way you can use either gesture, and you don't have to mess with the settings. But if you are developing an app where one of the gestures conflicts with something you are trying to capture in your app, you can disable one of the gestures (or even both!).

There is an advantage to offer two different gestures and letting you turn them on and off since there will almost always be some gesture you can use to get to the dev menu, even if you are adding shaking or three finger touching to your app for something.

I personally prefer the multifinger touch and would use that as my default way of doing things. Shaking the device is so time consuming and annoying.

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Fixes #7644 and brings us closer to the separate `expo-dev-menu` package that can live without kernel.
Also fixed https://twitter.com/bantingGamer/status/1251154340129062917

# How

- Added `EXDevMenuManager` singleton that manages the dev menu. Most parts of it were in kernel or some others classes strongly connected with kernel. Now, kernel is just its delegate.
- Refactored `EXKernelDevMotionHandler` to `EXDevMenuMotionInterceptor` which is much clearer and makes it possible to uninstall the interceptor (these interceptors are kinda hacky, I tried not to do them this way but well, RN does the same 🤷).
- Added `EXDevMenuGestureRecognizer` for long press gesture, with minimum press duration set to 0.5s and with light impact haptic feedback.
- Added `EXDevMenuGestureInterceptor` that swaps the default implementation of `UIView#gestureRecognizers`, but only for `UIWindow`s. This is needed to make three-finger gesture work everywhere in the app.
- Renamed `EXMenuViewController` to `EXDevMenuViewController` (all dev menu classes should be prefixed by `EXDevMenu`) and again, removed some connections to kernel.
- Added UI in user settings to select dev menu gestures (can select them separately).
- Fixed an issue with dev menu view being below modals in the view hierarchy.

# Test Plan

So far I've tested it only with published experience. I'm gonna do smoke tests on dev apps later.
I mostly focused on testing whether:
- the dev menu works in general (after refactoring)
- three-finger gesture works on the initial screen of the app, on the modals and on the dev menu itself
- all combinations of selected gestures work as expected
- selected gestures persist after killing and relaunching the client

### CODE REVIEW

**Comment by ccheever** [comment]
One other thing that I think we should add to this. When you first open up the app, there is a popup that says something like 

Hello there, friend! 

Since this is your first time opening the Expo app .... you can shake your device to get back to this menu at any time.

We should change that to say "Shake your device or long press anywhere on the screen with three fingers" or something like that.

I think this code is in `home/menu/DevMenuOnboarding.tsx`

File: home/screens/UserSettingsScreen.js [hunk 1]
```diff
@@ -113,6 +133,53 @@ export default class UserSettingsScreen extends React.Component {
...
+    const { devMenuSettings } = this.props;
+
+    if (!devMenuSettings || Platform.OS !== 'ios') {
+      return null;
+    }
+
+    return (
+      <View style={styles.sectionWrapper}>
+        <SectionLabelContainer>
+          <SectionLabelText>DEV MENU GESTURES</SectionLabelText>
```
Comment (brentvatne) [inline home/screens/UserSettingsScreen.js:146]:
```suggestion
          <SectionLabelText>DEVELOPER MENU GESTURES</SectionLabelText>
```

File: home/screens/UserSettingsScreen.js [hunk 2]
```diff
@@ -113,6 +133,53 @@ export default class UserSettingsScreen extends React.Component {
...
+            <GenericCardBody style={styles.cardBody}>
+              <GenericCardTitle>Three-finger long press</GenericCardTitle>
+            </GenericCardBody>
+            {devMenuSettings.touchGestureEnabled && this._renderCheckmark()}
+          </GenericCardContainer>
+        </TouchableHighlight>
+
+        <View style={SharedStyles.genericCardDescriptionContainer}>
+          <Text style={SharedStyles.genericCardDescriptionText}>
+            Selected gestures will toggle the dev menu while inside an experience. The menu allows
```
Comment (brentvatne) [inline home/screens/UserSettingsScreen.js:174]:
```suggestion
            Selected gestures will toggle the developer menu while inside an experience. The menu allows
```

**Review by brentvatne — APPROVED** [review]
this is great, excellent work @tsapeta!

**Pull request merged**